### PR TITLE
feat: add project, portfolio, and goal entity tools

### PR DESCRIPTION
**Pull request by rusakov-tech**
## Summary

- Adds full read/write MCP tool support for Yandex Tracker's **projects, portfolios, and goals** — `project_get`/`_find`/`_create`/`_update`/`_delete`, and the same set for `portfolio_*` and `goal_*` — backed by new protocol methods, client implementation, and Redis caching wrappers for the `/v3/entities` API. Tools are defined per entity type (not one generic "entity" tool), matching the granularity of the rest of the server.
- Adds **comment tools** for all three entity types: `*_get_comments`, `*_add_comment`, `*_update_comment`, `*_delete_comment` for project/portfolio/goal, hitting `v3/entities/<type>/<id>/comments`. Comments are deliberately excluded from caching since they mutate frequently.
- Adds `fields` selectors (enum-constrained) across entity, issue comment/worklog/attachment, and user tools so large payloads can be trimmed, matching the pattern already used for issues/queues.
- Fixes entity create/update calls to actually forward the `fields` query param (previously the response always came back with `fields: null`, forcing a follow-up `*_get`), splits `DEFAULT_ENTITY_FIELDS` per entity type (goals don't support `start`), and exposes `links`/relationships on project/portfolio/goal create and update so entities can be linked to each other.
- Adds LLM-facing guidance to the server's instructions and param descriptions: the queue-vs-entity distinction, how to attach an issue to a project (`project` field), how to nest projects/portfolios into a portfolio and goals under a parent goal (`parent_entity`), and how to set cross-entity relationships (`links`) — with concrete examples throughout.
- Clarifies tool/param descriptions for LLM callers (pagination defaults, `entityStatus` enum values, previously-undocumented required fields) and fixes a stale `issues_find` description and a redundant resolution hint on `issue_close`.

### CODE REVIEW

File: mcp_tracker/mcp/params.py [hunk 1]
```diff
@@ -102,14 +438,43 @@
...
+- Project/portfolio -> portfolio (containment, "included in portfolio"): set `parent_entity` on
+  `project_create`/`project_update`/`portfolio_create`/`portfolio_update` to
+  `{"primary": <portfolio id>, "secondary": [<other portfolio ids>]}`.
+- Goal -> parent goal (containment): set `parent_entity.primary` on `goal_create`/`goal_update` to the
+  parent goal's id.
+- Cross-entity relationships that aren't containment (e.g. a project "depends on" another project, or
+  "works towards" a goal): use the `links` param on `project_create`/`portfolio_create`/`goal_create`/
+  `*_update`, e.g. `[{"relationship": "works towards", "entity": <goal id>}]`. `links` REPLACES the
+  entity's full link set on update, so re-send existing links you want to keep alongside new ones -
+  fetch them first with `*_get(fields=["links"])` if unsure what's already there.
```
Comment (igorcoding) [inline mcp_tracker/mcp/params.py]: **Блокер: этот совет невыполним, а семантика `links` при этом деструктивная.**

`*_get(fields=["links"])` не сработает ни при каких условиях:

* `links` нет среди членов `ProjectFieldsEnum` / `PortfolioFieldsEnum` / `GoalFieldsEnum` — они генерируются из `*Fields.model_fields`, а поля `links` в этих моделях нет вообще (`grep -n links mcp_tracker/tracker/proto/types/entities.py` → пусто);
* значит MCP-валидация входа отвергнет `fields=["links"]` ещё до похода в API;
* и даже если бы дошло — `BaseTrackerEntity` объявлен с `extra="ignore"`, так что `links` из ответа молча выбросился бы.

В сумме получается худший из возможных вариантов: read-path для связей отсутствует, а write-path затирает набор целиком. Любой `project_update(links=[...])` безвозвратно сносит все существующие связи сущности, и восстановить их средствами самого сервера нечем — при том что инструкция выше явно обещает модели обратное и ведёт её в тупик.

Варианты:
1. добавить `links` в `ProjectFields`/`PortfolioFields`/`GoalFields` (и, соответственно, в enum) — тогда совет станет рабочим;
2. если API не отдаёт `links` через параметр `fields`, убрать невыполнимую рекомендацию и вместо неё поднять предупреждение: «`links` полностью заменяет набор; прочитать текущий набор через этот сервер нельзя».

Тот же текст про replace продублирован в `_LINKS_DESCRIPTION` (строка 244) — его тоже стоит поправить синхронно.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/mcp/tools/project_write.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+)
+from mcp_tracker.mcp.utils import get_yandex_auth
+from mcp_tracker.settings import Settings
+from mcp_tracker.tracker.proto.types.entities import ProjectEntity
+from mcp_tracker.tracker.proto.types.issues import IssueComment
+
+_QUEUE_RESTRICTIONS_NOTE = (
+    " Not subject to TRACKER_LIMIT_QUEUES / TRACKER_READ_ONLY_QUEUES restrictions, "
+    "since an entity isn't reliably mappable to a single queue."
+)
```
Comment (igorcoding) [inline mcp_tracker/mcp/tools/project_write.py]: **Блокер: обход ограничений по очередям нельзя выключить, и он шире, чем здесь написано.**

Само решение не спорю — сущность действительно не мапится на одну очередь, и честная оговорка лучше кривого маппинга. Проблема в том, что это текст в описании тулза, то есть подсказка модели, а не enforcement, и отказаться от новой поверхности оператор не может.

Что получается на практике:

1. **Регрессия для существующих деплоев.** Инсталляция с `TRACKER_LIMIT_QUEUES=DEV` после апгрейда молча получает org-wide чтение **и запись** проектов/портфелей/целей, включая `project_delete(with_board=True)`. Конфиг не менялся, гарантии изменились.
2. **Оговорка неполная.** Она висит только на write-тулзах, но `project.py` / `portfolio.py` / `goal.py` тоже не вызывают `check_*_access` — read-тулзы обходят ограничения ровно так же и об этом не сказано нигде. То же в `README.md:591`: «Entity **write** tools ... are also not currently subject to».
3. **Документация прямо противоречит коду.** Секция «Queue Access Control» (`README.md:1236-1256`) по-прежнему утверждает, что `TRACKER_LIMIT_QUEUES` действует «for both reads and writes», без каких-либо исключений. И там же рекламируется сценарий «shared MCP gateway where end users reach Tracker only through the server and never hold the raw token themselves» — то есть ровно тот деплой, где обход больнее всего.

Минимум, что стоит сделать:

* флаг включения группы — `TRACKER_ENTITIES_ENABLED` (по умолчанию `false`), либо авто-выключение entity-тулзов когда задан `TRACKER_LIMIT_QUEUES`, чтобы апгрейд не расширял периметр молча;
* перенести оговорку в секцию «Queue Access Control» обоих README и распространить её на чтение.

Бонусом флаг закрывает и проблему размера манифеста (см. комментарий в `tools/__init__.py`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/mcp/tools/__init__.py [hunk 3]
```diff
@@ -35,11 +47,17 @@ def register_all_tools(settings: Settings, mcp: FastMCP[Any]) -> None:
...
+    register_portfolio_tools(settings, mcp)
+    register_goal_tools(settings, mcp)
 
     # Only register write tools if not in read-only mode
     if not settings.tracker_read_only:
         register_queue_write_tools(settings, mcp)
         register_issue_write_tools(settings, mcp)
+        register_project_write_tools(settings, mcp)
+        register_portfolio_write_tools(settings, mcp)
+        register_goal_write_tools(settings, mcp)
```
Comment (igorcoding) [inline mcp_tracker/mcp/tools/__init__.py]: **Блокер: манифест тулзов вырастает в 3.2 раза — примерно +62k токенов в каждой сессии.**

Замерил фактический размер `await mcp.list_tools()` на обеих ветках (`description` + `inputSchema` + `outputSchema`, JSON):

| | `main` | эта ветка |
|---|---|---|
| Тулзов | 39 | 78 |
| Размер | 114 896 симв. (~29k токенов) | 364 295 симв. (~91k токенов) |
| Доля entity-тулзов | — | **68 %** (246 227 симв.) |

Разбивка по этой ветке: `descriptions` 13.5 KB, `inputSchema` 111 KB, **`outputSchema` 233 KB**. Основной вес — именно output-схемы: каждый write-тулз возвращает полную сущность, и `ProjectEntity` со всеми `$defs` инлайнится по ~6.3 KB, 39 раз подряд. Топ: `project_update` 12.2 KB, `project_create` 11.6 KB, `portfolio_update` 11.5 KB, `project_update_checklist` 9.3 KB.

~60k токенов накладных расходов на каждый диалог платят все пользователи сервера, включая тех, кто про проекты и цели никогда не спросит. Плюс часть MCP-клиентов имеет лимиты на число тулзов.

В описании PR выбор «per-entity, не generic entity-тулз» заявлен осознанно, и это твоё продуктовое решение — я его не оспариваю. Но при таком весе стоит хотя бы одно из:

* **флаг включения группы** (см. комментарий в `project_write.py`) — закрывает сразу две проблемы, и по умолчанию ничего не ломает;
* **объединить `project_*` и `portfolio_*`** — схемы идентичны с точностью до `issueQueues`, это 24 из 39 тулзов;
* **свернуть checklist-тулзы** — 6 штук на тип, 12 суммарно, ~90 KB; напрашивается один тулз с параметром операции.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/proto/types/entities.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    followers: list[UserReference] | None = NoneExcludedField
+    start: date | datetime | None = NoneExcludedField
+    end: date | datetime | None = NoneExcludedField
+    quarter: list[str] | None = NoneExcludedField
+    tags: list[str] | None = NoneExcludedField
+    entityStatus: ProjectPortfolioStatus | None = NoneExcludedField
+    parentEntity: ParentEntityRef | None = NoneExcludedField
+    teamAccess: bool | None = NoneExcludedField
+    issueQueues: list[QueueReference] | None = NoneExcludedField
+    lastCommentUpdatedAt: date | datetime | None = NoneExcludedField
```
Comment (igorcoding) [inline mcp_tracker/tracker/proto/types/entities.py]: **`date | datetime` молча теряет время и таймзону — и делает тип поля зависимым от данных.**

Порядок в юнионе здесь решает: pydantic пробует `date` первым, поэтому любой timestamp ровно в полночь схлопывается в голую дату вместе с offset'ом. Проверил на этой ветке:

```
'2024-01-01T00:00:00.000+0000' -> datetime.date(2024, 1, 1)       # сериализуется как "2024-01-01"
'2024-12-31T23:59:59.000+0000' -> datetime.datetime(..., tzinfo=TzInfo(0))  # "2024-12-31T23:59:59Z"
'2026-01-01'                   -> datetime.date(2026, 1, 1)
```

Yandex Tracker отдаёт `start`/`end` именно в формате `...T00:00:00.000+0000`, так что на практике offset теряется всегда. Для самих `start`/`end` это ещё терпимо — по докам это date-поля.

А вот `lastCommentUpdatedAt` (строка 91) — настоящий timestamp, и его сериализованный тип получается **недетерминированным**: `datetime` обычно, но `date` если комментарий пришёл ровно в полночь. Для потребителя-LLM это ровно те грабли, от которых `fields`-селекторы в этом PR и спасают.

Замечу, что `test_entity_field_dates.py:22` проходит только потому, что использует `23:59:59` — на `00:00:00` он бы поймал схлопывание.

Предлагаю: `lastCommentUpdatedAt: datetime` (и то же самое в `PortfolioFields:119`, `GoalFields:146`), а для `start`/`end` — либо тоже явный `datetime`, либо комментарий, что потеря offset сознательная. Смена порядка на `datetime | date` проблему не решает, а переворачивает: тогда `"2026-01-01"` станет наивным `datetime` в полночь.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/mcp/tools/goal_write.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+    @mcp.tool(
+        title="Delete Goal",
+        description="Delete a Yandex Tracker goal." + _QUEUE_RESTRICTIONS_NOTE,
+        annotations=ToolAnnotations(readOnlyHint=False, destructiveHint=True),
+    )
+    async def goal_delete(
+        ctx: Context[Any, AppContext],
+        entity_id: EntityID,
+        with_board: EntityWithBoardParam = False,
```
Comment (igorcoding) [inline mcp_tracker/mcp/tools/goal_write.py]: У целей в Трекере нет досок, так что `with_board` здесь бессмысленный — он только добавляет модели лишний параметр, о котором она будет гадать. У `project_delete` / `portfolio_delete` он на месте.

Предлагаю убрать его из `goal_delete` (и из `EntitiesProtocol.goal_delete` / `TrackerClient.goal_delete` соответственно), либо, если API его всё же принимает и молча игнорирует, оставить, но зафиксировать это в описании параметра.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/custom/client.py [hunk 6]
```diff
@@ -1010,3 +1028,1230 @@ async def issue_move(
...
+            fields_body["end"] = end.isoformat()
+        if tags is not None:
+            fields_body["tags"] = tags
+        if entity_status is not None:
+            fields_body["entityStatus"] = entity_status
+        if parent_entity is not None:
+            fields_body["parentEntity"] = parent_entity.model_dump(exclude_none=True)
+        if team_access is not None:
+            fields_body["teamAccess"] = team_access
+        return fields_body
```
Comment (igorcoding) [inline mcp_tracker/tracker/custom/client.py:1465]: Здесь всё построено на «пропускаем то, что `is None`», из чего следует: **очистить поле через `*_update` невозможно**. `tags=[]` сработает (пустой список — не `None`), а вот снять `description`, `lead` или `parent_entity` нечем.

Ограничение согласуется с тем, как устроен `issue_update`, так что менять поведение в этом PR не прошу — но стоит зафиксировать его в описаниях соответствующих параметров в `params.py` (`EntityDescriptionParam`, `EntityLeadParam`, `EntityParentEntityParam`), иначе модель будет пытаться передать `null` и молча получать no-op.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/custom/client.py [hunk 7]
```diff
@@ -1010,3 +1028,1230 @@ async def issue_move(
...
+        comment: str | None,
+        version: int | None,
+        fields: list[str] | None,
+        auth: YandexAuth | None,
+    ) -> bytes:
+        body: dict[str, Any] = {"fields": fields_body}
+        if links is not None:
+            body["links"] = [link.model_dump(exclude_none=True) for link in links]
+        if comment is not None:
+            body["comment"] = comment
```
Comment (igorcoding) [inline mcp_tracker/tracker/custom/client.py]: `fields_body` может оказаться пустым — например при `project_update(entity_id, links=[...])` или при `project_update(entity_id, comment="...")`, где ни одно поле из `_build_entity_fields_body` не задано. Тогда в API уходит `{"fields": {}, "links": [...]}`.

Скорее всего Трекер это проглотит, но кейс не покрыт ни одним тестом в `tests/tracker/custom/entities/test_entity_update.py`, а сценарий «поменять только связи» — вполне рабочий и, судя по инструкциям в `params.py`, ожидаемый от модели.

Предлагаю не класть `fields` в тело, когда `fields_body` пустой, и добавить на это тест. То же касается `_entity_create`, хотя там `summary` обязателен, так что пустым тело не будет.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/custom/client.py [hunk 8]
```diff
@@ -1010,3 +1028,1230 @@ async def issue_move(
...
+        entity_id: str,
+        *,
+        auth: YandexAuth | None,
+    ) -> list[IssueComment]:
+        async with self._session.get(
+            f"v3/entities/{entity_type}/{entity_id}/comments",
+            headers=await self._build_headers(auth),
+        ) as response:
+            response.raise_for_status()
+            return IssueCommentList.model_validate_json(await response.read()).root
```
Comment (igorcoding) [inline mcp_tracker/tracker/custom/client.py]: Пагинации нет — возвращается только первая страница, молча и без какого-либо признака, что данные обрезаны.

Ровно так же устроен существующий `issue_get_comments`, так что это не регрессия и на блокер не тяну. Но у проектов и портфелей обсуждения обычно длиннее, чем у отдельной задачи, а `*_get_comments` заявлен в описании тулза как «Get comments of a Yandex Tracker project» без оговорок — модель будет считать список полным.

Как минимум стоит завести issue; endpoint комментариев в v3 поддерживает `perPage` + курсор по `id`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by igorcoding — COMMENTED** [review]
## Ревью: projects, portfolios, goals

Проверял на локально вычекаченной ветке PR, с полным прогоном тулчейна.

### Статус проверок

| Проверка | Результат |
|---|---|
| `ruff check` / `ruff format --check` | ✅ чисто |
| `mypy .` | ✅ 154 файла, без ошибок |
| `ty check .` | ✅ чисто |
| `pytest` | ✅ 858 passed (583 на `main`, +275 тестов) |
| Аудит copy-paste | ✅ ни одного рассинхрона `project_*`→`portfolio_*`→`goal_*`, ни одного «потерянного» параметра в делегирующих обёртках (проверял AST-обходом всех слоёв) |

Механическое исполнение сильное: слои протокол → клиент → кеш → тулзы соблюдены, типизация строгая, тесты проверяют фактические HTTP-payload'ы через `RequestCapture`, документация обновлена на двух языках, `manifest.json` содержит ровно 78 записей и совпадает с реальным списком тулзов. Чек-лист из `AGENTS.md` пройден целиком.

### Блокеры

1. **`links` невозможно прочитать, но при `*_update` он затирается целиком** — инструкция в `params.py:477` предлагает модели `*_get(fields=["links"])`, чего enum полей не допускает, а модели `*Fields` этого поля вообще не содержат. Получается деструктивная запись без read-path. → комментарий в `mcp_tracker/mcp/params.py`
2. **Entity-тулзы обходят `TRACKER_LIMIT_QUEUES` / `TRACKER_READ_ONLY_QUEUES`, и отключить их нельзя** — существующие деплои после апгрейда молча получают org-wide запись, включая `project_delete(with_board=True)`; секция «Queue Access Control» в README при этом продолжает утверждать обратное. → комментарий в `mcp_tracker/mcp/tools/project_write.py`
3. **Манифест тулзов растёт в 3.2 раза: 39 → 78 тулзов, 115 KB → 364 KB схем (~+62k токенов в каждой сессии)**, 68 % веса — новые entity-тулзы, из них основная масса приходится на output-схемы. → комментарий в `mcp_tracker/mcp/tools/__init__.py`

Пункты 2 и 3 закрываются одним флагом включения группы.

### Замечания

4. `date | datetime` схлопывает полуночные timestamp'ы в голую дату вместе с таймзоной; для `lastCommentUpdatedAt` это делает тип поля зависящим от данных → `mcp_tracker/tracker/proto/types/entities.py`
5. `with_board` у `goal_delete` — у целей нет досок → `mcp_tracker/mcp/tools/goal_write.py`
6. `_entity_update` может отправить `{"fields": {}}` при апдейте только связей или только комментария; кейс не покрыт тестом → `mcp_tracker/tracker/custom/client.py`
7. Поля нельзя очистить через `*_update` — стоит зафиксировать в описаниях параметров → `mcp_tracker/tracker/custom/client.py`
8. `*_get_comments` без пагинации, возвращает первую страницу молча → `mcp_tracker/tracker/custom/client.py`

### Итог

Работа большая и качественная, к реализации как таковой претензий почти нет. От мержа держат три пункта выше, причём два из трёх — это одно решение про флаг, а третий (`links`) требует либо расширить модели, либо переписать инструкцию так, чтобы она не вела модель в тупик.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/proto/types/entities.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+    Readable through the `fields` selector; this server does not write key
+    results. `progressPercentage` on the goal is derived from these.
+    """
+
+    id: str | None = NoneExcludedField
+    text: str | None = NoneExcludedField
+    # How the result is measured: "value" (quantitative) or "binary" (done/not done).
+    type: str | None = NoneExcludedField
+    deadline: ChecklistItemDeadline | None = NoneExcludedField
```
Comment (igorcoding) [inline mcp_tracker/tracker/proto/types/entities.py]: **Новое (из `1817a4d`): переиспользование `ChecklistItemDeadline` может уронить весь `goal_get`.**

У `ChecklistItemDeadline` (`types/issues.py`) обязательны все три поля:

```python
date: datetime.datetime
deadline_type: str = Field(validation_alias=...)   # required
is_exceeded: bool = Field(validation_alias=...)    # required
```

Остальные новые модели (`EntityMetricItem`, `GoalKeyResultProgress`, сам `GoalKeyResultItem`) сделаны полностью опциональными — а `deadline` выбивается. Если Трекер вернёт у key result дедлайн без `isExceeded`, падает валидация **всей сущности**, а не одного поля:

```
GoalFields.model_validate({'keyResultItems': [
    {'id': '1', 'text': 'KR', 'type': 'value',
     'deadline': {'date': '2026-08-20T00:00:00.000+0000', 'deadlineType': 'date'}}]})

ValidationError: keyResultItems.0.deadline.isExceeded — Field required
```

То есть `goal_get(fields=[..., "keyResultItems"])` вернёт ошибку целиком, вместо того чтобы отдать цель без дедлайна. Тест `test_entity_field_dates.py:97` проверяет только happy path, где `isExceeded` присутствует.

Косвенный аргумент, что форма может отличаться: пример в `EntityChecklistItemDeadlineParam` (`params.py`) — это `{'date': ..., 'deadlineType': 'date'}`, без `isExceeded`.

Предлагаю завести отдельный `EntityDeadline` с опциональными `deadline_type` / `is_exceeded` (в духе остальных новых моделей) и использовать его в `GoalKeyResultItem`.

Оговорюсь: у `checklistItems` в `ProjectFields`/`PortfolioFields` ровно та же хрупкость через `ChecklistItem.deadline`, и она появилась ещё в ранних коммитах этого PR — я её в прошлый заход пропустил. Если правишь, имеет смысл закрыть оба места сразу.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/custom/client.py [hunk 10]
```diff
@@ -1010,3 +1060,1259 @@ async def issue_move(
...
+        async with self._session.get(
+            f"v3/entities/{entity_type}/{entity_id}/comments/_relative",
+            headers=await self._build_headers(auth),
+            params=params,
+        ) as response:
+            response.raise_for_status()
+            page = EntityCommentsRelativePage.model_validate_json(await response.read())
+            next_cursor: str | None = None
+            if page.hasNext and page.comments:
+                next_cursor = str(page.comments[-1].id)
```
Comment (igorcoding) [inline mcp_tracker/tracker/custom/client.py]: Логика курсора хорошая — в частности, `page.hasNext and page.comments` корректно закрывает случай «`hasNext: true`, но список пуст», и на это есть отдельный тест (`test_no_next_cursor_on_empty_last_page`). Спасибо, что покрыл.

Один вопрос, который тестами не проверяется, потому что они на моках: курсор строится как `str(comments[-1].id)` и уходит в `from`, но у `IssueComment` есть и `long_id`. Если `_relative` ожидает в `from` длинный идентификатор, пагинация молча сломается на втором вызове против живого API, а тесты останутся зелёными.

Проверял ли на реальной организации, что вторая страница действительно приходит? Если да — вопрос снят, стоит только добавить строчку в докстринг `_entity_get_comments`, что `from` принимает короткий числовой `id`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/tracker/proto/types/issues.py [hunk 11]
```diff
@@ -244,6 +262,17 @@ class ChangelogEntry(CreatedUpdatedMixin, BaseTrackerEntity):
     )
 
 
+class CommentsPage(BaseTrackerEntity):
+    """A page of comments plus the cursor to fetch the next page.
+
+    `next_cursor` is parsed from the `Link: rel="next"` response header; it is `None`
+    when there are no more pages. Pass it back as the `cursor` argument to continue.
+    """
```
Comment (igorcoding) [inline mcp_tracker/tracker/proto/types/issues.py:287]: Мелочь: докстринг теперь верен только наполовину. Для `issue_get_comments` `next_cursor` действительно берётся из заголовка `Link: rel="next"`, а вот для entity-комментариев он выводится из `hasNext` + `id` последнего комментария в `_relative` — заголовка `Link` там нет.

Модель общая для обоих случаев, так что предлагаю переформулировать нейтрально: «`next_cursor` — курсор следующей страницы или `None`, если страниц больше нет; способ его получения зависит от эндпоинта».

---
_Generated by [Claude Code](https://claude.ai/code)_

File: mcp_tracker/mcp/tools/issue_read.py [hunk 12]
```diff
@@ -65,20 +69,40 @@ async def issue_get(
...
+        cursor: CommentsCursorParam = None,
+        fields: Annotated[
+            list[CommentFieldsEnum] | None,
+            Field(
+                description="Fields to include in each comment. In order to not pollute the context "
+                "window - select only the fields you need (comment text/text_html can be large). "
+                "Not specifying this returns all available fields.",
+            ),
+        ] = None,
+    ) -> CommentsPage:
```
Comment (igorcoding) [inline mcp_tracker/mcp/tools/issue_read.py:102]: Смена типа возврата правильная и консистентна с `issue_get_changelog`, но обращу внимание: это **ломающее изменение контракта уже существующего инструмента**, и оно, в отличие от entity-тулзов, не спрятано за `TRACKER_ENTITIES_ENABLED`.

`issue_get_comments` был `list[IssueComment]`, стал `CommentsPage` (`{comments, next_cursor}`). Любой существующий пользователь сервера, который обходил результат как массив, сломается после апдейта — независимо от того, включил он сущности или нет. Плюс поведение по умолчанию меняется: раньше приходили все комментарии, теперь первая страница из 50.

Просьба не откатывать, а просто отразить это явно — в описании PR (в саммари сейчас этого пункта нет) и в release notes при выпуске версии, чтобы попало в changelog как breaking.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by igorcoding — COMMENTED** [review]
## Повторное ревью после правок (`490833a`…`835ba47`)

Перепроверил все восемь замечаний на новом HEAD — **закрыты все восемь**, причём несколько сделаны основательнее, чем я просил. Спасибо.

### Статус проверок

| Проверка | Результат |
|---|---|
| `ruff check` / `ruff format --check` | ✅ чисто |
| `mypy .` / `ty check .` | ✅ 155 файлов, без ошибок |
| `pytest` | ✅ **1006 passed** (было 858) |
| Аудит copy-paste и проброса параметров | ✅ 0 расхождений |

### По замечаниям

| # | Замечание | Как закрыто |
|---|---|---|
| 1 | `links` без read-path | Проверено по api-ref: links write-only. Невыполнимая инструкция убрана, вместо неё явный WARNING + требование подтвердить полный набор у пользователя; в модели оставлен комментарий, почему поля нет |
| 2 | Обход queue-ограничений | `TRACKER_ENTITIES_ENABLED` (default `false`), гейт на регистрации read и write; `ENTITY_QUEUE_RESTRICTIONS_NOTE` теперь на всех тулзах, включая read; секции Queue Access Control в обоих README поправлены + `AGENTS.md` |
| 3 | Размер манифеста | По умолчанию **39 тулзов / 119 KB** (на `main` — 39 / 115 KB), с флагом — 78 / 396 KB. Дефолтный профиль фактически вернулся к базовому |
| 4 | `date \| datetime` | `lastCommentUpdatedAt: datetime` во всех трёх моделях + комментарий, почему для `start`/`end` юнион оставлен |
| 5 | `with_board` у целей | Убран из тулза, протокола и клиента |
| 6 | Пустой `{"fields": {}}` | Не отправляется, в `_entity_create` и `_entity_update`; есть тесты |
| 7 | Нельзя очистить поле | `_CANNOT_CLEAR_NOTE` на `description` / `lead` / `parent_entity` |
| 8 | Пагинация комментариев | Курсорная пагинация через `_relative` для сущностей и через `Link`-заголовок для задач, с тестом на edge-case `hasNext` + пустой список |

Отдельно отмечу `tests/mcp/server/test_manifest.py` — тест, который пришивает `manifest.json` к фактически зарегистрированным тулзам и проверяет пометку opt-in. Это ровно тот класс расхождений, который я ловил руками; хорошо, что теперь ловится в CI. Проводка `TRACKER_ENTITIES_ENABLED` в `user_config` MCPB и в `resources.py` тоже к месту.

### Новое

Появился `1817a4d` с `metricItems` / `keyResultItems` — расширение скоупа уже в ходе ревью, но сделано аккуратно и честно задокументировано как read-only с объяснением, почему запись не поддержана. Одно замечание по нему:

* **`GoalKeyResultItem.deadline` переиспользует строгий `ChecklistItemDeadline`** — при отсутствии `isExceeded` в ответе падает валидация всей сущности, а не одного поля. Единственное блокирующее из нового. → комментарий в `proto/types/entities.py`

И три некритичных:

* курсор entity-комментариев строится как `str(comments[-1].id)`, тесты на моках это не проверяют — вопрос, тестировалось ли против живой организации → `custom/client.py`
* докстринг `CommentsPage` описывает только issue-вариант получения курсора → `proto/types/issues.py`
* смена типа возврата `issue_get_comments` — ломающее изменение существующего инструмента, не закрытое флагом; просьба отразить в описании PR и release notes → `mcp/tools/issue_read.py`

### Итог

Все прошлые блокеры сняты. Остаётся один по существу — строгий `deadline` у key results; остальное косметика и один вопрос про реальный API.

---
_Generated by [Claude Code](https://claude.ai/code)_
